### PR TITLE
Feature/147 get min data

### LINKED ISSUE #147: [Online Learning] Modules has to specify how much past data they need.
Some modules need multiple value to calculate the current value. E.g. rolling_mean and sample_module. However, the current online/batch learning approach only provides the past batch as input, which might not be enough. 

## Solution:
* [ ] The modules need a method which returns the number of needed data.
* [ ] The step has to pay attention that enough data is fetched from the previous step.

see #116

### PR DESCRIPTION
**Pull request by benHeid**
## Related Issues
closes #147 
## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos)
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [x] Did you update the CHANGELOG

## PR review
Anyone in the community is free to review the PR once the tests have passed.

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones to the PR so it can be classified

### CODE REVIEW

File: pywatts/modules/preprocessing/clock_shift.py [hunk 1]
```diff
@@ -75,3 +77,6 @@ def transform(self, x: xr.DataArray) -> xr.DataArray:
                 f"Not all indexes ({indexes}) are in the indexes of x ({list(x.indexes.keys())}).",
                 "Perhaps you set the wrong indexes with set_params or during the initialization of the ClockShift.",
                 module=self.name) from exc
+
+    def get_min_data(self):
```
Comment (SMEISEN) [inline pywatts/modules/preprocessing/clock_shift.py:81]: please add docstrings

File: pywatts/modules/feature_extraction/trend_extraction.py [hunk 2]
```diff
@@ -83,3 +83,6 @@ def transform(self, x: xr.DataArray) -> xr.DataArray:
                   range(1, self.length + 1)]
         trend = xr.DataArray(np.stack(trends, axis=-1), dims=(*x.dims, "length"), coords=x.coords)
         return trend.transpose(_get_time_indexes(x)[0], "length", ...)
+
+    def get_min_data(self):
```
Comment (SMEISEN) [inline pywatts/modules/feature_extraction/trend_extraction.py:87]: please add docstrings

File: pywatts/core/base_step.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def __init__(self, input_steps: Optional[Dict[str, "BaseStep"]] = None,
         self.transform_time = SummaryObjectList(self.name + " Transform Time", category=SummaryCategory.TransformTime)
 
     def get_result(self, start: pd.Timestamp, end: Optional[pd.Timestamp], buffer_element: str = None,
-                   return_all=False):
+                   return_all=False, minimum_data=(0, pd.Timedelta(0))):
```
Comment (SMEISEN) [inline pywatts/core/base_step.py:52]: add docsrings for `minimum_data` (param & type)

File: pywatts/core/step.py [hunk 4]
```diff
@@ -144,15 +145,15 @@ def load(cls, stored_step: Dict, inputs, targets, module, file_manager):
 
         return step
 
-    def _compute(self, start, end):
-        input_data = self._get_input(start, end)
-        target = self._get_target(start, end)
+    def _compute(self, start, end, minimum_data):
```
Comment (SMEISEN) [inline pywatts/core/step.py:148]: please add docstrings

File: pywatts/core/step.py [hunk 5]
```diff
@@ -173,14 +174,26 @@ def _compute(self, start, end):
...
+        else:
+            minimum_data = minimum_data[0], minimum_data[1] + min_data_module
         return {
-            key: target.get_result(start, batch) for key, target in self.targets.items()
+            key: target.get_result(start, batch, minimum_data=minimum_data)
+            for key, target in self.targets.items()
         }
 
-    def _get_input(self, start, batch):
+    def _get_input(self, start, batch, minimum_data=(0, pd.Timedelta(0))):
```
Comment (SMEISEN) [inline pywatts/core/step.py:188]: please add docstrings

File: pywatts/core/base_step.py [hunk 6]
```diff
@@ -118,12 +119,16 @@ def further_elements(self, counter: pd.Timestamp) -> bool:
                 return False
         return True
 
-    def _pack_data(self, start, end, buffer_element=None, return_all=False):
+    def _pack_data(self, start, end, buffer_element=None, return_all=False, minimum_data=(0, pd.Timedelta(0))):
```
Comment (SMEISEN) [inline pywatts/core/base_step.py:122]: please add docstrings

File: pywatts/core/step.py [hunk 7]
```diff
@@ -173,14 +174,26 @@ def _compute(self, start, end):
             result_dict[key] = res.sel(**{_get_time_indexes(res)[0]: index[(index >= start)]})
         return result_dict
 
-    def _get_target(self, start, batch):
+    def _get_target(self, start, batch, minimum_data=(0, pd.Timedelta(0))):
+        min_data_module = self.module.get_min_data()
+        if isinstance(min_data_module, (int, np.integer)):
+            minimum_data = minimum_data[0] + min_data_module, minimum_data[1]
+        else:
```
Comment (SMEISEN) [inline pywatts/core/step.py:181]: is it possible that `min_data_module` is an `int` and the argument `minimum_data` is a `pd.Timedelta`, i.e. both if and else should be executed?

File: pywatts/core/base_step.py [hunk 8]
```diff
@@ -74,10 +74,11 @@ def get_result(self, start: pd.Timestamp, end: Optional[pd.Timestamp], buffer_el
         # Only execute the module if the step is not finished and the results are not yet calculated
         if not self.finished and not (end is not None and self._current_end is not None and end <= self._current_end):
             if not self.buffer or not self._current_end or end > self._current_end:
-                self.cached_result["cached"] = self._compute(start, end)
+                self.cached_result["cached"] = self._compute(start, end, minimum_data)
                 self.cached_result["start"] = start
                 self.cached_result["end"] = end
                 self._current_end = end
+                self.minimum_data = minimum_data
```
Comment (SMEISEN) [inline pywatts/core/base_step.py:81]: this is a new argument that is never used, should it be `self.cached_result["minimum_data"]`?

File: pywatts/core/base_step.py [hunk 9]
```diff
@@ -89,13 +90,13 @@ def get_result(self, start: pd.Timestamp, end: Optional[pd.Timestamp], buffer_el
 
         # Check if the cached results fits to the request, if yes return it.
         if self.cached_result["cached"] is not None and self.cached_result["start"] == start and self.cached_result[
-            "end"] == end:
+            "end"] == end and self.cached_result["minimum_data"] == minimum_data:
```
Comment (SMEISEN) [inline pywatts/core/base_step.py:93]: `self.cached_result["minimum_data"]` gets initialized but never reassigned, i.e. it is always `None`

File: pywatts/core/step.py [hunk 10]
```diff
@@ -173,14 +174,26 @@ def _compute(self, start, end):
...
+            key: target.get_result(start, batch, minimum_data=minimum_data)
+            for key, target in self.targets.items()
         }
 
-    def _get_input(self, start, batch):
+    def _get_input(self, start, batch, minimum_data=(0, pd.Timedelta(0))):
+        min_data_module = self.module.get_min_data()
+        if isinstance(min_data_module, (int, np.integer)):
+            minimum_data = minimum_data[0] + min_data_module, minimum_data[1]
+        else:
```
Comment (SMEISEN) [inline pywatts/core/step.py:192]: see commen line 181

**Review by SMEISEN — CHANGES_REQUESTED** [review]
Thank you for your contribution, please note my comments.

File: pywatts/core/step.py (hunk 7 above)
Comment (benHeid) [reply pywatts/core/step.py:181]: No, this should not be allowed. The get_min_data in the modules should only return int or timedelta. Mixing this in one module would be confusing I suppose.

File: pywatts/core/step.py (hunk 10 above)
Comment (benHeid) [reply pywatts/core/step.py:192]: See remark above.

**Pull request merged**